memory/allocator.go

```diff
@@ -170,6 +213,28 @@ func (a *Allocator) allocator() memory.Allocator {
 	return a.Allocator
 }
 
+// Manager will manage the memory allowed for the Allocator.
+// The Allocator may use the Manager to request additional memory or to
+// give back memory that is currently in use by the Allocator
+// when/if it is no longer needed.
+type Manager interface {
+	// RequestMemory will request that the given amount of memory
+	// be reserved for the caller. The manager will return the number
+	// of bytes that were successfully reserved. The n value will be
+	// either equal to or greater than the requested number of bytes.
+	// If the manager cannot reserve at least bytes in memory, then
+	// it will return an error with the details.
+	RequestMemory(want int64) (got int64, err error)
+
+	// FreeMemory will inform the memory manager that this memory
+	// is not being used anymore.
+	// It is not required for this to be called similar to how
+	// it is not necessary for a program to free the memory.
+	// It is the responsibility of the controller to reclaim the
+	// unfreed memory when a resource is dead.
+	FreeMemory(bytes int64)
```

Comment:
This part seems a bit odd. The Manager is not aware of the controller, so how can it delegate freeing to the controller?

Should we update this interface with a `Release` method to make this more explicit?

Reply:
I can change the wording. The intention was to say that it is the responsibility of the memory manager itself to reclaim its memory. A release method wouldn't really help as I would be unlikely to implement it.

The original reason I put this here was in case it ever did get used, I wouldn't make a breaking interface change. And it's not a problem if the underlying implementation doesn't implement it because memory is managed by the underlying resource manager anyway.

I don't think release would have a meaning for this interface.

Reply:
Updated the text to remove mention of the controller and to make clear that this is purely an optional method. It is the responsibility of the manager to identify when the resource is dead and to reclaim the memory itself similar to how an OS would reclaim memory from a dead program.